### PR TITLE
Add backend and UI support for direct joins between tables

### PR DESCRIPTION
**Pull request by mosiac1**
### Describe your changes :

Currently, table joins can only exist between columns, which doesn't cover all cases. This allows joins to exist directly between two tables.  On the UI these appear only in "Frequently Joined Tables", but for a specific column like column joins.

#
### Type of change :

- [x] New feature

#
### Frontend Preview (Screenshots) :
Everything looks the same, changes to the frontend are just to merge direct table joins and column joins.

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

Frontend: @open-metadata/ui
Backend: @open-metadata/backend

### CODE REVIEW

**Review by harshach — APPROVED** [review]

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java [hunk 1]
```diff
@@ -779,12 +780,38 @@ List<List<String>> listFromByAllPrefix(
...
+        "SELECT fromFQN, toFQN, json FROM field_relationship WHERE "
+            + "fromFQN = :fqn AND fromType = :type AND toType = :otherType AND relation = :relation "
+            + "UNION "
+            + "SELECT toFQN, fromFQN, json FROM field_relationship WHERE "
+            + "toFQN = :fqn AND toType = :type AND fromType = :otherType AND relation = :relation")
+    @RegisterRowMapper(ToFieldMapper.class)
+    List<Triple<String, String, String>> listBidirectional(
+        @Bind("fqn") String fqn,
+        @Bind("type") String type,
+        @Bind("otherType") String otherType,
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java:799]: @mosiac1 what is this otherType supposed to store? any additional information we need can be stored here https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/sql/com.mysql.cj.jdbc.Driver/v001__create_db_connection_info.sql#L30

Comment (mosiac1) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java:799]: it's not storing any new information, it's just referring to this [column](https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/sql/com.mysql.cj.jdbc.Driver/v001__create_db_connection_info.sql#L27)

`listBidirectional` get all field relations where one side has fqn = `fqn` and type = `type` and the other has type = `otherType`

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java [hunk 2]
```diff
@@ -779,12 +780,38 @@ List<List<String>> listFromByAllPrefix(
...
 
+    @SqlQuery(
+        "SELECT fromFQN, toFQN, json FROM field_relationship WHERE "
+            + "fromFQN = :fqn AND fromType = :type AND toType = :otherType AND relation = :relation "
+            + "UNION "
+            + "SELECT toFQN, fromFQN, json FROM field_relationship WHERE "
+            + "toFQN = :fqn AND toType = :type AND fromType = :otherType AND relation = :relation")
+    @RegisterRowMapper(ToFieldMapper.class)
+    List<Triple<String, String, String>> listBidirectional(
+        @Bind("fqn") String fqn,
```
Comment (sureshms) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java:797]: We have EntityRepository.findBoth() method that can be replaced with this method call.

**Comment by mosiac1** [comment]
> @mosiac1, can you create an issue for this PR or add more details on what you mean by...

Sorry, tought I had added an issue. 

https://github.com/open-metadata/OpenMetadata/issues/5041

This discribes the bigger picture that this PR fits into and the goal I'm trying to reach - creating more metadata from usage.

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java (hunk 2 above)
Comment (mosiac1) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/CollectionDAO.java:797]: I think` EntityRepository.findBoth()` is for looking up EntityRelationships, not FieldRelationships. I can add similar querries for EntityRelationships, but I think that should go in another PR

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/TableRepository.java [hunk 3]
```diff
@@ -157,11 +168,20 @@ public Table addJoins(UUID tableId, TableJoins joins) throws IOException {
       validateColumnFQNs(join.getJoinedWith());
     }
 
+    // Validate direct table joins
+    for (JoinedWith join : joins.getDirectTableJoins()) {
+      validateTableFQN(join.getFullyQualifiedName());
+    }
+
```
[the remark this answers is not in the document]
Comment (mosiac1) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/TableRepository.java:175]: Well, the columns from `joins.getColumnJoins().getJoinedWith()` are not necessarily from tables in `joins.getDirectTableJoins()` and vice-versa. 

This is on me for not explaining very well what this PR is adding.  You can think of  `joins.getDirectTableJoins()` as storing joins between the this tables and others tables where there are no clear columns used for the join.

File: catalog-rest-service/src/main/resources/json/schema/entity/data/table.json [hunk 4]
```diff
@@ -242,6 +242,20 @@
       "required": ["name", "dataType"],
       "additionalProperties": false
     },
+    "joinedWith": {
+      "description": "Fully qualified names of the columns that this column is joined with.",
+      "type": "object",
```
[the remark this answers is not in the document]
Comment (mosiac1) [reply catalog-rest-service/src/main/resources/json/schema/entity/data/table.json:243]: Updated 👍

**Comment by mosiac1** [comment]
@ayush-shah I'm having problems with generating python types from the schemas (I made some changes to the schema, but these shouldn't be breaking anything). It may be related to the refactoring you are working on for the python packages. Could you please take a look?

**Comment by ayush-shah** [comment]
Yes, looking into it.

**Review by sureshms — APPROVED** [review]

**Pull request merged**